File: windows/Hinge.Core/NetworkInterfaceHelper.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net;
using System.Net.NetworkInformation;
using System.Net.Sockets;

namespace Hinge.Core;

public sealed record PhysicalLanEndpoint(
    NetworkInterface Interface,
    IPAddress Address,
    IPAddress Mask,
    IPAddress Broadcast);

public static class NetworkInterfaceHelper
{
    private static readonly string[] VirtualOrVpnKeywords =
    {
        "vpn", "tap", "tun", "wintun", "wireguard", "openvpn",
        "cisco", "anyconnect", "fortinet", "forticlient", "globalprotect",
        "sangfor", "easyconnect", "tailscale", "zerotier", "softether",
        "clash", "sing-box", "shadowsocks", "v2ray", "meta tunnel",
        "nexgen", "hyper-v", "vmware", "virtualbox", "virtual",
        "pseudo", "npcap", "teredo", "6to4", "bluetooth", "wan miniport",
        "qemu", "veth", "docker", "wsl"
    };

    /// <summary>
    /// Checks whether the given network interface is a virtual adapter, VPN tunnel,
    /// or host-only virtual switch rather than a genuine physical LAN/Wi-Fi adapter.
    /// </summary>
    public static bool IsVirtualOrVpnInterface(NetworkInterface nic)
    {
        if (nic == null) return true;
        if (nic.OperationalStatus != OperationalStatus.Up) return true;

        if (nic.NetworkInterfaceType is NetworkInterfaceType.Loopback or
            NetworkInterfaceType.Tunnel or
            NetworkInterfaceType.Ppp or
            NetworkInterfaceType.Slip)
        {
            return true;
        }

        string name = nic.Name ?? string.Empty;
        string desc = nic.Description ?? string.Empty;

        foreach (var keyword in VirtualOrVpnKeywords)
        {
            if (name.IndexOf(keyword, StringComparison.OrdinalIgnoreCase) >= 0 ||
                desc.IndexOf(keyword, StringComparison.OrdinalIgnoreCase) >= 0)
            {
                return true;
            }
        }

        return false;
    }

    /// <summary>
    /// Checks whether an IPv4 address belongs to a virtual, loopback, APIPA, or RFC 2544 benchmark range.
    /// </summary>
    public static bool IsReservedOrVirtualIp(IPAddress address)
    {
        if (address.AddressFamily != AddressFamily.InterNetwork) return true;
        byte[] bytes = address.GetAddressBytes();

        // 127.0.0.0/8 (Loopback)
        if (bytes[0] == 127) return true;

        // 0.0.0.0/8 (Current network)
        if (bytes[0] == 0) return true;

        // 169.254.0.0/16 (Link-local / APIPA)
        if (bytes[0] == 169 && bytes[1] == 254) return true;

        // 198.18.0.0/15 (RFC 2544 benchmark, standard fake-ip range used by Clash / sing-box / tun2socks)
        if (bytes[0] == 198 && (bytes[1] == 18 || bytes[1] == 19)) return true;

        return false;
    }

    /// <summary>
    /// Calculates the directed broadcast address for a given IPv4 address and subnet mask.
    /// </summary>
    public static IPAddress CalculateBroadcastAddress(IPAddress address, IPAddress mask)
    {
        byte[] ipBytes = address.GetAddressBytes();
        byte[] maskBytes = mask.GetAddressBytes();
        var broadcast = new byte[4];
        for (int i = 0; i < 4; i++)
        {
            broadcast[i] = (byte)(ipBytes[i] | (byte)~maskBytes[i]);
        }
        return new IPAddress(broadcast);
    }

    /// <summary>
    /// Discovers all active physical LAN and Wi-Fi endpoints on this machine.
    /// </summary>
    public static IReadOnlyList<PhysicalLanEndpoint> GetPhysicalLanEndpoints()
    {
        var endpoints = new List<PhysicalLanEndpoint>();

        foreach (var nic in NetworkInterface.GetAllNetworkInterfaces())
        {
            if (IsVirtualOrVpnInterface(nic)) continue;

            IPInterfaceProperties properties;
            try
            {
                properties = nic.GetIPProperties();
            }
            catch
            {
                continue;
            }

            foreach (var unicast in properties.UnicastAddresses)
            {
                if (unicast.Address.AddressFamily != AddressFamily.InterNetwork) continue;
                if (IsReservedOrVirtualIp(unicast.Address)) continue;

                var mask = unicast.IPv4Mask ?? IPAddress.Parse("255.255.255.0");
                var broadcast = CalculateBroadcastAddress(unicast.Address, mask);

                endpoints.Add(new PhysicalLanEndpoint(nic, unicast.Address, mask, broadcast));
            }
        }

        return endpoints;
    }

    /// <summary>
    /// Returns a list of all local physical IPv4 addresses as strings.
    /// </summary>
    public static List<string> GetPhysicalCandidateAddresses()
    {
        return GetPhysicalLanEndpoints()
            .Select(ep => ep.Address.ToString())
            .Distinct()
            .ToList();
    }

    /// <summary>
    /// Finds a local physical endpoint that shares the same subnet with the target remote IP.
    /// </summary>
    public static PhysicalLanEndpoint? FindMatchingPhysicalLanEndpoint(IPAddress targetIp)
    {
        if (targetIp.AddressFamily != AddressFamily.InterNetwork) return null;
        byte[] targetBytes = targetIp.GetAddressBytes();

        var endpoints = GetPhysicalLanEndpoints();
        foreach (var ep in endpoints)
        {
            byte[] localBytes = ep.Address.GetAddressBytes();
            byte[] maskBytes = ep.Mask.GetAddressBytes();

            bool sameSubnet = true;
            for (int i = 0; i < 4; i++)
            {
                if ((targetBytes[i] & maskBytes[i]) != (localBytes[i] & maskBytes[i]))
                {
                    sameSubnet = false;
                    break;
                }
            }

            if (sameSubnet)
            {
                return ep;
            }
        }

        return null;
    }

    /// <summary>
    /// Finds a local physical IP that shares the same subnet with the target remote IP.
    /// Used to explicitly bind sockets before connecting or sending probes.
    /// </summary>
    public static IPAddress? FindMatchingLocalPhysicalAddress(IPAddress targetIp)
    {
        var matchingEp = FindMatchingPhysicalLanEndpoint(targetIp);
        if (matchingEp != null) return matchingEp.Address;

        if (!IPAddress.IsLoopback(targetIp))
        {
            var endpoints = GetPhysicalLanEndpoints();
            if (endpoints.Count > 0) return endpoints[0].Address;
        }

        return null;
    }
}

File: android/lib/core/network_interface_helper.dart
import 'dart:io';

class PhysicalInterfaceEndpoint {
  final NetworkInterface networkInterface;
  final InternetAddress address;
  final InternetAddress broadcast;
  final int prefixLength;

  const PhysicalInterfaceEndpoint({
    required this.networkInterface,
    required this.address,
    required this.broadcast,
    required this.prefixLength,
  });
}

class NetworkInterfaceHelper {
  static const List<String> _virtualKeywords = [
    'tun',
    'tap',
    'ppp',
    'p2p',
    'dummy',
    'lo',
    'rmnet',
    'sit',
    'ip6tnl',
    'wsl',
    'docker',
    'veth',
    'vpn',
    'wireguard',
    'wintun',
    'clash',
    'singbox',
    'shadowsocks',
    'tailscale',
    'zerotier',
    'meta',
    'nexgen',
  ];

  /// Checks whether an interface name suggests it is a virtual or VPN tunnel.
  static bool isVirtualOrVpnInterfaceName(String name) {
    final lower = name.toLowerCase();
    for (final keyword in _virtualKeywords) {
      if (lower.contains(keyword)) {
        return true;
      }
    }
    return false;
  }

  /// Checks whether an IPv4 address is in loopback, APIPA, or RFC 2544 benchmark ranges.
  static bool isReservedOrVirtualIp(InternetAddress address) {
    if (address.type != InternetAddressType.IPv4) return true;
    final parts = address.address.split('.');
    if (parts.length != 4) return true;
    final b0 = int.tryParse(parts[0]) ?? 0;
    final b1 = int.tryParse(parts[1]) ?? 0;

    // Loopback (127.0.0.0/8)
    if (b0 == 127) return true;
    // Current network (0.0.0.0/8)
    if (b0 == 0) return true;
    // APIPA / Link-local (169.254.0.0/16)
    if (b0 == 169 && b1 == 254) return true;
    // RFC 2544 benchmark (198.18.0.0/15, used by Clash/sing-box/tun2socks fake-IP)
    if (b0 == 198 && (b1 == 18 || b1 == 19)) return true;

    return false;
  }

  /// Safely calculates the directed broadcast address for an IPv4 address and prefix length.
  static InternetAddress calculateBroadcastAddress(
    InternetAddress address, [
    int prefixLength = 24,
  ]) {
    final parts = address.address.split('.');
    if (parts.length != 4) return InternetAddress('255.255.255.255');
    final octets = parts.map((p) => int.tryParse(p) ?? 0).toList();
    final prefix = prefixLength.clamp(8, 30);

    final ip =
        (octets[0] << 24) | (octets[1] << 16) | (octets[2] << 8) | octets[3];
    final mask = (0xffffffff << (32 - prefix)) & 0xffffffff;
    final broadcastInt = (ip | (~mask & 0xffffffff)) & 0xffffffff;

    final b0 = (broadcastInt >> 24) & 0xff;
    final b1 = (broadcastInt >> 16) & 0xff;
    final b2 = (broadcastInt >> 8) & 0xff;
    final b3 = broadcastInt & 0xff;

    return InternetAddress('$b0.$b1.$b2.$b3');
  }

  /// Discovers all physical LAN/Wi-Fi endpoints on this device.
  static Future<List<PhysicalInterfaceEndpoint>>
  getPhysicalLanEndpoints() async {
    final endpoints = <PhysicalInterfaceEndpoint>[];
    try {
      final interfaces = await NetworkInterface.list(
        type: InternetAddressType.IPv4,
        includeLoopback: false,
        includeLinkLocal: false,
      );

      for (final nic in interfaces) {
        if (isVirtualOrVpnInterfaceName(nic.name)) continue;

        for (final addr in nic.addresses) {
          if (isReservedOrVirtualIp(addr)) continue;

          var prefix = 24;
          InternetAddress? directedBroadcast;

          try {
            final dynamic dynamicAddr = addr;
            final dynamic p = dynamicAddr.prefixLength;
            if (p is int && p > 0 && p <= 32) {
              prefix = p;
            }
            final dynamic b = dynamicAddr.broadcast;
            if (b is InternetAddress) {
              directedBroadcast = b;
            }
          } catch (_) {
            // Getter might not be present on all Dart platforms
          }

          directedBroadcast ??= calculateBroadcastAddress(addr, prefix);

          endpoints.add(
            PhysicalInterfaceEndpoint(
              networkInterface: nic,
              address: addr,
              broadcast: directedBroadcast,
              prefixLength: prefix,
            ),
          );
        }
      }
    } catch (_) {
      // Network listing failure fallback
    }

    return endpoints;
  }

  /// Returns candidate physical IPv4 address strings.
  static Future<List<String>> getPhysicalCandidateAddresses() async {
    final endpoints = await getPhysicalLanEndpoints();
    return endpoints.map((e) => e.address.address).toSet().toList();
  }

  /// Finds a local physical IP that shares the same subnet with the target IP.
  static Future<InternetAddress?> findMatchingLocalPhysicalAddress(
    InternetAddress targetIp,
  ) async {
    if (targetIp.type != InternetAddressType.IPv4) return null;
    final targetParts = targetIp.address.split('.');
    if (targetParts.length != 4) return null;
    final t0 = int.tryParse(targetParts[0]) ?? 0;
    final t1 = int.tryParse(targetParts[1]) ?? 0;
    final t2 = int.tryParse(targetParts[2]) ?? 0;
    final t3 = int.tryParse(targetParts[3]) ?? 0;
    final targetInt = (t0 << 24) | (t1 << 16) | (t2 << 8) | t3;

    final endpoints = await getPhysicalLanEndpoints();
    for (final ep in endpoints) {
      final localParts = ep.address.address.split('.');
      if (localParts.length != 4) continue;
      final l0 = int.tryParse(localParts[0]) ?? 0;
      final l1 = int.tryParse(localParts[1]) ?? 0;
      final l2 = int.tryParse(localParts[2]) ?? 0;
      final l3 = int.tryParse(localParts[3]) ?? 0;
      final localInt = (l0 << 24) | (l1 << 16) | (l2 << 8) | l3;

      final prefix = ep.prefixLength.clamp(8, 30);
      final mask = (0xffffffff << (32 - prefix)) & 0xffffffff;

      if ((targetInt & mask) == (localInt & mask)) {
        return ep.address;
      }
    }

    if (endpoints.isNotEmpty && !targetIp.isLoopback) {
      return endpoints.first.address;
    }

    return null;
  }
}
